### PR TITLE
Code Docs

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
r? @M-J-Hooper  

bors d=M-J-Hooper

File: crates/hir/src/semantics.rs [hunk 1]
```diff
@@ -697,6 +697,25 @@ fn find_root(node: &SyntaxNode) -> SyntaxNode {
...
+///
+/// It is a bit tricky, as scopes do not really exist inside the compiler.
+/// Rather, the compiler directly computes for each reference the definition it
+/// refers to. It might transiently compute the explicit scope map while doing
+/// so, but, generally, this is not something left after the analysis.
+///
+/// However, we do very much need explicit scopes for IDE purposes --
+/// completion, at its core, lists the contents of the current scope. Notion of
+/// scope is also useful to answer question like "what would be the meaning of
+/// this piece of code if we insert into this position?".
```
Comment (lnicola) [inline crates/hir/src/semantics.rs:711]:
```suggestion
/// completion, at its core, lists the contents of the current scope. The notion
/// of scope is also useful to answer questions like "what would be the meaning
/// of this piece of code if we inserted it into this position?".
```

File: crates/hir/src/semantics.rs [hunk 2]
```diff
@@ -697,6 +697,25 @@ fn find_root(node: &SyntaxNode) -> SyntaxNode {
...
+///
+/// So `SemanticsScope` is constructed from a specific program point (a syntax
+/// node or just a raw offset) and provides access to the set of visible names
+/// in on a somewhat best-effort basis.
```
Comment (lnicola) [inline crates/hir/src/semantics.rs:715]:
```suggestion
/// on a somewhat best-effort basis.
```

File: crates/hir/src/semantics.rs [hunk 3]
```diff
@@ -697,6 +697,25 @@ fn find_root(node: &SyntaxNode) -> SyntaxNode {
...
+///
+/// Note that if you are wondering "what this specific existing name means?",
```
Comment (lnicola) [inline crates/hir/src/semantics.rs:717]:
```suggestion
/// Note that if you are wondering "what does this specific existing name mean?",
```

**Review by M-J-Hooper — APPROVED** [review]
Thanks for adding this! This kind of documentation is really valuable to someone like me who is quite new to rust and the ra project in general (and only has limited time to dip in and out every once in a while).

**Pull request merged**